src/java.base/share/classes/jdk/internal/foreign/abi/CapturableState.java

```diff
@@ -29,67 +29,67 @@
 import java.lang.foreign.StructLayout;
-import java.lang.foreign.ValueLayout;
-import java.util.List;
-import java.util.stream.Collectors;
-import java.util.stream.Stream;
+import java.util.ArrayList;
+import java.util.Map;
 
 import static java.lang.foreign.ValueLayout.JAVA_INT;
 
-public enum CapturableState {
-    GET_LAST_ERROR    ("GetLastError",    JAVA_INT, 1 << 0, OperatingSystem.isWindows()),
-    WSA_GET_LAST_ERROR("WSAGetLastError", JAVA_INT, 1 << 1, OperatingSystem.isWindows()),
-    ERRNO             ("errno",           JAVA_INT, 1 << 2, true);
+/**
+ * Utility class for the call states to capture.
+ */
+public final class CapturableState {
 
-    public static final StructLayout LAYOUT = MemoryLayout.structLayout(
-        supportedStates().map(CapturableState::layout).toArray(MemoryLayout[]::new));
-    public static final List<CapturableState> BY_ORDINAL = List.of(values());
+    public static final StructLayout LAYOUT;
+    private static final Map<String, Integer> MASKS;
 
     static {
-        assert (BY_ORDINAL.size() < Integer.SIZE); // Update LinkerOptions.CaptureCallState
-    }
-
-    private final String stateName;
-    private final ValueLayout layout;
-    private final int mask;
-    private final boolean isSupported;
-
-    CapturableState(String stateName, ValueLayout layout, int mask, boolean isSupported) {
-        this.stateName = stateName;
-        this.layout = layout.withName(stateName);
-        this.mask = mask;
-        this.isSupported = isSupported;
-    }
-
-    private static Stream<CapturableState> supportedStates() {
-        return Stream.of(values()).filter(CapturableState::isSupported);
-    }
-
-    public static CapturableState forName(String name) {
-        return Stream.of(values())
-                .filter(stl -> stl.stateName().equals(name))
-                .filter(CapturableState::isSupported)
-                .findAny()
-                .orElseThrow(() -> new IllegalArgumentException(
-                        "Unknown name: " + name +", must be one of: "
-                            + supportedStates()
-                                    .map(CapturableState::stateName)
-                                    .collect(Collectors.joining(", "))));
-    }
-
-    public String stateName() {
-        return stateName;
+        if (OperatingSystem.isWindows()) {
+            LAYOUT = MemoryLayout.structLayout(
+                    JAVA_INT.withName("GetLastError"),
+                    JAVA_INT.withName("WSAGetLastError"),
+                    JAVA_INT.withName("errno"));
+            MASKS = Map.of(
+                    "GetLastError",    1 << 0,
+                    "WSAGetLastError", 1 << 1,
+                    "errno",           1 << 2
+            );
+        } else {
+            LAYOUT = MemoryLayout.structLayout(
+                    JAVA_INT.withName("errno"));
+            MASKS = Map.of(
+                    "errno",           1 << 2
+            );
+        }
```

Comment:
Pre-existing, but while you're here, could you add a comment stating that the mask values have to be kept in sync with the value in `DowncallLinker::capture_state` in `/hotspot/share/prims/downcallLinker.cpp`? (See https://github.com/openjdk/jdk/blob/master/src/hotspot/share/prims/downcallLinker.cpp#L36)